### ISSUE TITLE
General configuration when building gcc

### ISSUE THREAD

**Issue by th-otto** [freemint/m68k-atari-mint-gcc#28]
Hi,

i just wanted to summarize some configuration options that i've used in the past when configuring gcc, and explain why they were choosen. There are also some additional tasks that should be performed in the build script. It would be nice if we can agree to use a consistent concept so that compilers build from different sources behave the same.

- --target="${TARGET}" --build="$BUILD"
The obvious setting to build a cross compiler. ``$TARGET`` here is either m68k-atari-mint or m68k-atari-mintelf. ``$BUILD`` is the canonical system name of the host, something like x86_64-pc-linux or x86_64-pc-cygwin.

- --prefix="${PREFIX}"
The main directory prefix where the compiler should be installed. Until now, this was ``/usr``. For the elf toolchain, this is currently ``/opt/cross-mintelf``. Once we decide that the compiler is mature enough, that should imho be changed to ``/usr`` again (all toolchain dependant file are stored in ``m68k-atari-mintelf`` subdirectories, anyway, so they don't conflict with a.out toolchains)

- --libdir="$PREFIX/lib64" --bindir="${PREFIX}/bin" --libexecdir='${libdir}'
These are the settings that i use on my system, but those may be distribution dependant.

- --with-pkgversion="$REVISION"
The method i choosed to get the build version/date into the output of ``gcc -v``. Imho easier than to patch the sources.

- --disable-libcc1 --disable-werror
Maybe not strictly needed, but should not hurt.

- --with-sysroot="${PREFIX}/${TARGET}/sys-root" 
This is rather important, as it has an influence where the compiler will search by default for include headers. Without this, installing the toolchain into a different directory than it was configured for may not work. The header files are also used when configuring the c++ library, and thus have an influence about the generated ``c++config.h``.

- --with-gcc-major-version-only 
This affects the name of the subdirectories where actual executables like cc1 are placed, and also the default subdirectory of c++ headers. I want to use only ``13`` here, not ``13.2.0``, as i don't want to have different directories for each and every gcc version that was ever released. This seems to be also done by most distributions for the host gcc.

- --with-gxx-include-dir=${PREFIX}/${TARGET}/sys-root/usr/include/c++/13
Since headers for the c-library will be looked up in ``<sys-root>``, c++ headers are placed there, too.

- $gcc4_compat 
This was --with-default-libstdcxx-abi=gcc4-compatible until now, but we agreed to drop that for gcc >= 13

- --disable-libgomp
libgomp requires posix threads, which we don't have

- --without-newlib
We use the headers from mintlib, not newlib

- --disable-libstdcxx-pch
Disables installation of precompiled include files

- --disable-threads
We don't have posix conforming threads until now

- --disable-win32-registry
Prevents gcc registering itself on cygwin/mingw

- $enable_lto $enable_plugin
This two are both enabled for elf toolchains, and disabled otherwise.

- --enable-ssp --enable-libssp
Not sure whether those are strictly needed

- --disable-decimal-float
This will disable the new decimal float types. We don't have them in fdlibm, so enabling that would build a rather large support library that was never needed until now.

- --disable-nls
Maybe just to make it consistent with a native compiler, where we don't have i18n support. Also i don't think that anybody will ever update the *.po files for new translations from prgmint.c/elf32-atariprg.c

- $with_dw2_exceptions
Experimental support for dwarf2 exceptions. Does not work yet.

- --enable-languages="c,c++"
Minimum set of languages to build.

- --with-libstdcxx-zoneinfo=/usr/share/zoneinfo
This avoids compiling in a rather large static version of the hosts tzdata.zi.

To support multiple toolchains of different major compiler versions, some additional steps need to be done in the script. For example, the libstdc++.a is placed into ``<prefix>/<target>/lib`` by default. I move them to the directory where libgcc.a is also located, usually something like ``<libdir>/<target>/gcc/13``, so they won't conflict. Same also applies to any *.spec files.

For the elf toolchains, i also place symlinks to the lto plugin into the directory where the linker will look for them. For this to work, the symlinks are named something like ``liblto_plugin_mintelf.so.13`` so there are no conflicts with system libraries, and also not for different compiler versions.

Edit: --disable-werror seems to be absolutely needed, otherwise compilation might fail if your hosts compiler is newer, and suddenly spits out new warnings. A good example for this is when trying to compile the old gcc-4 codebase ;)

**Comment by mikrosk** [comment]
My set is very similar, I just omit a few of them because they seem to be autodetected (and/or I copied it from Vincent's scripts):

		--prefix=${INSTALL_DIR}
		--target=${TARGET} \
		--with-sysroot=${INSTALL_DIR}/${TARGET}/sys-root \
		--disable-nls \
		--enable-languages="c,c++" \
		--disable-libstdcxx-pch \
		--disable-libgomp \
		--with-cpu=${CPU} \
		--disable-lto \
		--with-libstdcxx-zoneinfo=no

For native builds it is also very important to specify these two separately:

		--with-sysroot="/" \
		--with-build-sysroot="${INSTALL_DIR}/${TARGET}/sys-root" \

`--with-cpu` not only specifies default CPU when invoking gcc without any switches but also influences the value of `CPU` in https://github.com/freemint/m68k-atari-mint-gcc/blob/gcc-13-mintelf/gcc/config/m68k/t-mint and therefore how multilib directories are constructed in https://github.com/freemint/m68k-atari-mint-gcc/blob/gcc-13-mintelf/gcc/config/m68k/t-mlibs.

**Comment by vinriviere** [comment]
Good summary.

For memory, I use those settings for my Ubuntu builds:
https://github.com/vinriviere/cross-mintelf-ubuntu/blob/8c9e4d2ee7d58340317a44fae6657ce56f85968f/gcc-m68k-atari-mintelf/debian/rules#L45-L54

Most important flaw is that I don't use `--with-sysroot`. But I will do.

**Comment by vinriviere** [comment]
BTW, I'm experiencing hardware failure. Building GCC repeatedly killed my laptop's SSD, probably because overheat. I will need a few days to recover.
NB: I delete this out-of-topic post right now.

**Comment by vinriviere** [comment]
BTW, @th-otto, do you really have a directory named `sys-root` somewhere on the filesystem? Containing something before building gcc? Sorry, but that's really mysterious.

**Comment by th-otto** [comment]
Yes, that directory exists and must contain atleast the mintlib headers and libraries when you try to build the c++ backend, otherwise the configure script will early stop because the ``C compiler cannot create executables``. It basically is the root of a filesystem that you would install natively, and is later populated also with all other libraries.

Its mysterious because i don't a have separate script to bootstrap gcc. That means you have to install mintlib there before building gcc. But i think that is also the case in your ubuntu scripts.

**Comment by mikrosk** [comment]
Yeah, one of the not-so-intuitive changes is to install all of your software (mintlib, gemlib etc) into `<path>/m68k-atari-mintelf/sys-root/usr` instead of just `<path>/m68k-atari-mintelf`. Since some time (years ago...) this is a defacto standard for gcc people, so we have just adopted it.

It has its advantages -- you can basically map (via smb, nfs etc) your platform's sys-root to your build machine (and vice versa).

**Comment by th-otto** [comment]
>you can basically map (via smb, nfs etc) your platform's sys-root to your build machine (and vice versa).

Theoretically, yes. But remember that this is a real root directory, and therefor also contains directories like ``/etc``. Files in that directory may be overwritten by ``make install``, thus overwriting your already customized configuration files.

Apropos nfs: i personally don't need it on atari, but our version only supports nfsv2. On a lot of distros, that is not supported anymore. Might be time to update it.

**Comment by th-otto** [comment]
There is another problem when replacing the compiler on native systems. When cross-compiling, everything is stored under some ``m68k-atari-mint`` or ``m68k-atari-mintelf`` prefix. But not on a native system. Header files should not be a problem, as they ideally should be identical. But libraries are, we only have one ``/usr/lib`` and ``/usr/local/lib`` directory. Any idea how to solve that?

**Comment by mikrosk** [comment]
As I person who provides native builds... I wouldn't worry about that too much. I've always viewed the archive as "from scratch installation", i.e. nobody ever asked me how to upgrade the compiler, it was assumed that every new gcc requires a clean slate.

So here we can say the same. Whoever is brave enough to support more than one toolchain natively (!), he can always install stuff into some /opt or /usr/local subdirectory and use -L as appropriate...

**Comment by vinriviere** [comment]
> But libraries are, we only have one `/usr/lib` and `/usr/local/lib` directory. Any idea how to solve that?

If multiple compilers produce incompatible binaries, you must use different nonstandard directories for the libs. This is already the case for multilibs.

IMHO, the key is to always use gcc's `--print-file-name` option to find the right directories.

Examples on Cygwin (cross-compiler, I admit):
```
$ m68k-atari-mintelf-gcc --print-file-name=libgcc.a
/opt/cross-mintelf/lib/gcc/m68k-atari-mintelf/13.2.0/libgcc.a

$ m68k-atari-mintelf-gcc -print-libgcc-file-name
/opt/cross-mintelf/lib/gcc/m68k-atari-mintelf/13.2.0/libgcc.a

$ m68k-atari-mintelf-gcc --print-file-name=libc.a
/opt/cross-mintelf/lib/gcc/m68k-atari-mintelf/13.2.0/../../../../m68k-atari-mintelf/lib/libc.a

$ m68k-atari-mintelf-gcc -print-prog-name=as
/opt/cross-mintelf/lib/gcc/m68k-atari-mintelf/13.2.0/../../../../m68k-atari-mintelf/bin/as.exe

$ echo $(dirname $(realpath $(m68k-atari-mintelf-gcc --print-file-name=libc.a)))
/opt/cross-mintelf/m68k-atari-mintelf/lib

$ CC=m68k-atari-mintelf-gcc
$ LIBDIR=$(dirname $(realpath $($CC --print-file-name=libc.a)))
$ echo $LIBDIR
/opt/cross-mintelf/m68k-atari-mintelf/lib
```

So as long as you know the name of the gcc executable, you know the right libdir. Even if it's is a nonstandard place.

**Comment by vinriviere** [comment]
> Its mysterious because i don't a have separate script to bootstrap gcc. That means you have to install mintlib there before building gcc. But i think that is also the case in your ubuntu scripts.

I finally used `--with-sysroot` in my official Ubuntu builds. Not in a clean way, but works well to produce expected binaries. Basically, I still install an older version of mintlib/fdlibm on the system. Then I create a sys-root folder, and inside I create symlinks to actual include/lib directories. Hackish, but works like a charm.

https://github.com/vinriviere/cross-mintelf-ubuntu/blob/4a6abc8a1e7ecdd2556b781dd205cdd113602baf/gcc-m68k-atari-mintelf/debian/rules#L29-L32
https://github.com/vinriviere/cross-mintelf-ubuntu/blob/4a6abc8a1e7ecdd2556b781dd205cdd113602baf/gcc-m68k-atari-mintelf/debian/rules#L54

Now there is no more `trap #7` inside my libgcc/unwind-dw2.o. Instead of that, there are calls to `abort()`, as expected.

**Comment by th-otto** [comment]
Just for info: i've just pushed a backported commit from gcc-7 to the gcc-4_6-mint branch. That commit will make gcc-4.6.4 now behave the same as newer compilers when looking up C++ headers when configured with --sys-root. Without this, if you installed the binaries to a directory different than the configured prefix, but have the headers installed in /usr/m68k-atari-mint/sys-root, the c++ compiler won't find its include files. For C-header files, this already worked, but not for c++.

**Comment by th-otto** [comment]
> `--with-cpu` not only specifies default CPU when invoking gcc without any switches but also influences the value of `CPU` in https://github.com/freemint/m68k-atari-mint-gcc/blob/gcc-13-mintelf/gcc/config/m68k/t-mint and therefore how multilib directories are constructed in https://github.com/freemint/m68k-atari-mint-gcc/blob/gcc-13-mintelf/gcc/config/m68k/t-mlibs.

That seems to be the case only with your scripts. In my installation, i compile the native compiler with `--with-default-cpu=xxx`. That has the effect that you always wanted, compile for m68020 by default when installing the compiler that was itself compiled for 020. Same should be the case for coldfire, but i cannot test that. But it keeps the directories in the same layout as for the cross-compiler (m68000 being in /usr/lib). But it works:
```
$ echo "" | gcc -dM -E - | grep 680
#define __mc68020__ 1
#define __mc68000__ 1
#define __mc68030__ 1
#define __mc68000 1
#define __mc68020 1
#define __mc68030 1
#define __mc68060__ 1
#define __M68020__ 1
#define __mc68040 1
#define __mc68060 1
#define mc68000 1
#define mc68020 1
#define mc68030 1
#define mc68040 1
#define mc68060 1
#define __mc68040__ 1
```

```
$ gcc --print-file-name=libgcc.a
/c/usr/bin/../lib/gcc/m68k-atari-mint/13/m68020-60/libgcc.a
```
So no need to mess around with different directory layout for different compilers.

**Comment by mikrosk** [comment]
> That seems to be the case only with your scripts. In my installation, i compile the native compiler with --with-default-cpu=xxx

Are you sure this is the flag? I can't find any trace of it neither in gcc nor online.

**Comment by th-otto** [comment]
Ah, i meant `--with-cpu`. But the result is the same.
